lib/galaxy/datatypes/binary.py

```diff
@@ -981,12 +981,15 @@ def set_meta(self, dataset, overwrite=True, **kwd):
         super(Anndata, self).set_meta(dataset, overwrite=overwrite, **kwd)
         try:
             with h5py.File(dataset.file_name, 'r') as anndata_file:
+                ## NOTE: It appears that 'dataset.metadata' variables can
+                ##       only be set exactly once, so tread carefully.
```

Comment:
```suggestion
```
That does not seem right.

Reply:
Hmm, so maybe my conclusion  was wrong but I am definitely seeing weird behaviour if add 

`dataset.metadata.shape = util.unicodify(anndata_file.attrs.get('shape'))`  on L995

The error message I get is:

Traceback (most recent call last):
  File "/home/tetris/repos/_galaxy/galaxy-upstream/lib/galaxy/jobs/__init__.py", line 1547, in _finish_dataset
    dataset.set_peek(line_count=line_count)
  File "/home/tetris/repos/_galaxy/galaxy-upstream/lib/galaxy/model/__init__.py", line 2779, in set_peek
    return self.datatype.set_peek(self, **kwd)
TypeError: set_peek() got an unexpected keyword argument 'line_count'

During handling of the above exception, another exception occurred:

```
Traceback (most recent call last):
  File "/home/nope/repos/_galaxy/galaxy-upstream/lib/galaxy/jobs/runners/__init__.py", line 540, in _finish_or_resubmit_job
    job_wrapper.finish(tool_stdout, tool_stderr, exit_code, check_output_detected_state=check_output_detected_state, job_stdout=job_stdout, job_stderr=job_stderr)
  File "/home/nope/repos/_galaxy/galaxy-upstream/lib/galaxy/jobs/__init__.py", line 1681, in finish
    self._finish_dataset(
  File "/home/nope/repos/_galaxy/galaxy-upstream/lib/galaxy/jobs/__init__.py", line 1550, in _finish_dataset
    dataset.set_peek()
  File "/home/nope/repos/_galaxy/galaxy-upstream/lib/galaxy/model/__init__.py", line 2779, in set_peek
    return self.datatype.set_peek(self, **kwd)
  File "/home/nope/repos/_galaxy/galaxy-upstream/lib/galaxy/datatypes/binary.py", line 1077, in set_peek
    peekstr = "[n_obs x n_vars]\n    %d x %d" % tuple(tmp.shape)
TypeError: 'NoneType' object is not iterable
FAIL
```

So I read this as:

 * `dataset.metadata.shape = util.unicodify(anndata_file.attrs.get('shape'))`  evaluates to `None` 
 * The `if 'X' in dataset.metadata.layers_names:` block lower down is either never reached, or the `dataset.metadata.shape` variable is never overwritten.

Some more tests:

* Evaluating `util.unicodify(anndata_file.attrs.get('shape'))` does not produce any errors, so it's the assignment to `dataset.metadata.shape` that seems to  cause the problem
* Surrounding `dataset.metadata.shape = util.unicodify(anndata_file.attrs.get('shape'))` in a try-catch block does not fix the issue, so I think the if-else block at the bottom *is* being reached.

Reply:
Can you remove the try/except catching on line 984 / 1058 ?
From what you describe I think the most likely explanation is that you're raising an exception, and the set_peek code relies on no exception having occurred.

That said I don't see `dataset.metadata.shape = util.unicodify(anndata_file.attrs.get('shape'))` on line 995, has something gone wrong there ? ` File "/home/nope/repos/_galaxy/galaxy-upstream/lib/galaxy/datatypes/binary.py", line 1077, in set_peek` also doesn't match the line numbers here.

Reply:
oh I meant "if you add <code snippet> at this line, this is what you get"

I can push this as a commit if you like, but I'd rather leave it out of the git history

Reply:
We can always rebase this out.

Reply:
> Can you remove the try/except catching on line 984 / 1058 ?

We should keep this though, right? I took this template from one of the other h5 classes in the file (Loom)